### PR TITLE
scanner: clean up

### PR DESCRIPTION
**Pull request by novakne**
**Todo:**

- [x] use std.fmt to define custom Formatter for title/camel case
- [x] replace `printIdentifier()` and `isValidIdentifier()` with std.zig.fmtId
- [x] Use the original variable name instead of title_case / snake_case
- [x] add zig-out/ to .gitignore

Maybe more thing?

### CODE REVIEW

**Comment by novakne** [comment]
I thought about doing something like this, but it add a lot of functions and still doesn't works perfectly
https://zigbin.io/520e43

**Comment by novakne** [comment]
I made some errors in the commit message, I'll fix this after everything is done

**Comment by novakne** [comment]
@ifreund Unless you want me to change some things, I'm done with what I wanted to do I think

**Comment by novakne** [comment]
rebase and squash commits

**Comment by novakne** [comment]
Just removed a forgotten useless 'prefix' in enum Case

File: src/scanner.zig [hunk 1]
```diff
@@ -484,11 +493,11 @@ const Interface = struct {
...
         try writer.print(
             \\ = struct {{
-            \\ extern const {s}_interface: common.Interface;
+            \\ extern const {[in]}_interface: common.Interface;
             \\ pub fn getInterface() callconv(.Inline) *const common.Interface {{
-            \\  return &{s}_interface;
+            \\  return &{[in]}_interface;
             \\ }}
-        , .{ interface.name, interface.name });
+        , .{ .in = fmtId(interface.name) });
```
Comment (ifreund) [inline src/scanner.zig:500]: this shouldn't use `fmtId()` as the produced code would be invalid if interface.name was an identifier in the current scope:
```zig
extern const @"foobar"_interface
```

File: src/scanner.zig [hunk 2]
```diff
@@ -667,14 +664,13 @@ const Message = struct {
...
                             try writer.print(
                                 \\ )) {{
-                                \\    .Enum => @intCast({s}, @enumToInt(_{s})),
-                                \\    .Struct => @bitCast(u32, _{s}),
+                                \\    .Enum => @intCast({[ct]}, @enumToInt(_{[an]})),
+                                \\    .Struct => @bitCast(u32, _{[an]}),
                                 \\    else => unreachable,
                                 \\ }}
-                            , .{ c_type, arg.name, arg.name });
+                            , .{ .ct = fmtId(c_type), .an = fmtId(arg.name) });
```
Comment (ifreund) [inline src/scanner.zig:671]:
```suggestion
                            , .{ .ct = c_type, .an = fmtId(arg.name) });
```

**Review by ifreund — CHANGES_REQUESTED** [review]
Overall this is looking like a very nice improvement, thanks.

What would you think about using longer names for the format specifiers? For example `type`/`@"type"` instead of `tc`, `namespace` instead of `ns`, `interface` instead of `in`, etc.

**Comment by novakne** [comment]
> What would you think about using longer names for the format specifiers? For example `type`/`@"type"` instead of `tc`, `namespace` instead of `ns`, `interface` instead of `in`, etc.

I could change it yeah, I hesitated to use longer names, it would be more clear but also more verbose, we could have some really long lines, so would it make it really more clear I'm not sure. I guess it depends on each one preference

**Comment by novakne** [comment]
hmm also `@"type"` gonna messed up a lot of things

for exemple 
``` zig
// works here
        try writer.print(
            \\pub const {[@"type"]} = opaque {{
            \\ pub const getInterface = common.{[namespace]}.{[interface]}.getInterface;
        , .{
            .@"type" = titleCaseTrim(interface.name),

// but not here:
        for (interface.enums.items) |e| {
            try writer.print("pub const {[@"type"]} = common.{[namespace]}.{[interface]}.{[@"type"]};\n", .{
                .@"type" = titleCase(e.name),

```
It could be a bit messy

**Comment by ifreund** [comment]
> hmm also `@"type"` gonna messed up a lot of things

using the `@""` syntax inside the format specifier isn't needed and probably won't work.

```zig
            try writer.print("pub const {[type]} = common.{[namespace]}.{[interface]}.{[@"type"]};\n", .{
                .@"type" = titleCase(e.name),
```

**Comment by novakne** [comment]
Let me now if that what you want and if it looks better with long names for you

**Review by ifreund — COMMENTED** [review]
I like the longer names, they make the format strings much more readable IMO

File: src/scanner.zig [hunk 3]
```diff
@@ -369,20 +365,28 @@ const Interface = struct {
...
-                    \\    return @ptrCast(*server.wl.Resource, _{s}).{s}();
+                    \\pub fn {[function0]}(_{[interface]}: *{[type]}) {[function1]} {{
+                    \\    return @ptrCast(*server.wl.Resource, _{[interface]}).{[function0]}();
                     \\}}
-                , .{ func[0], snake_case, title_case, func[1], snake_case, func[0] });
+                , .{
+                    .function0 = camelCase(func[0]),
+                    .interface = fmtId(trimPrefix(interface.name)),
+                    .@"type" = titleCaseTrim(interface.name),
+                    .function1 = camelCase(func[1]),
```
[the remark this answers is not in the document]
Comment (novakne) [reply src/scanner.zig:375]: how about `function_name` or just `function` for `function0` then?

[the remark this answers is not in the document]
Comment (ifreund) [reply src/scanner.zig:375]: > how about `function_name` or just `function` for `function0` then?

just function sounds good to me!

**Comment by novakne** [comment]
Well it should all be good now, want me to squash all of this in one commit?

**Comment by ifreund** [comment]
Thanks! I just went ahead and squashed it myself.

**Pull request merged**